File: 142. Linked List Cycle II/_142.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:
        slow = head
        fast = head
        flag = 0

        while fast.next and fast.next.next:
            fast = fast.next.next
            slow = slow.next

            if fast == slow:
                flag = 1
                break

        if flag == 0:
            return None

        while head != slow:
            head = head.next
            slow = slow.next

        return head

File: 142. Linked List Cycle II/142.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

struct ListNode {
    int val;
    struct ListNode *next;
};

struct ListNode *detectCycle(struct ListNode *head) {
    struct ListNode* fast = head;
    struct ListNode* slow = head;
    int flag = 0;

    while(fast != NULL && fast->next != NULL){
        fast = fast->next->next;
        slow = slow->next;

        if (fast == slow){
            flag = 1;
            break;
        }
    }

    if (flag == 0){
        return NULL;
    }

    while(head != slow){
        head = head->next;
        slow = slow->next;
    }

    return head;
}